### PR TITLE
feat: Storage cache get flag

### PR DESCRIPTION
**Pull request by markphelps**
Fixes: FLI-540

This removes caching `GetFlag` at the server/GRPC interceptor layer and instead moves it down to the storage layer so that all calls to `GetFlag` in storage can take advantage of the cached value (ie: both boolean and variant evaluation)

⚠️ It is a departure however from the previous caching strategy laid out in our [docs](https://www.flipt.io/docs/configuration/storage#caching).

Now basically we only rely on TTL to purge the cache instead of trying to delete flags from cache whenever the flag (or its children such as variants) changes.

This will have repercussions in the UI for example if someone has cache enabled and then modifies a flag they will not immediately see the actual value, but will instead see the old value. 

I think we can mitigate this in one of two ways (or both)
- [x] provide support for [Cache-Control headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) in the backend, and have the frontend pass in headers saying basically 'we don't ever want the cached value'. This should keep the UI consistent with the state in the db regardless of cache
- [ ] update our caching and production docs to state:
  1. the updated caching policy (only read through, dependent on TTL for invalidation)
  2. update our [HA/production docs](https://www.flipt.io/docs/operations/deployment#further-considerations)  to mention you may not want to enable cache for the write instance, and only enable it for the read instances

🐛  Also fixes an insidious bug where the cacher was not getting initialized for the middleware because of Go shadowing 

Will have the UI use the new `Cache-Control` header in follow up PR

## Load Testing

~~Here are some basic load tests showing the difference between cached and non-cached, note that there isn't much difference in the `Boolean` evaluation case, so I need to investigate further but wanted to get this PR up for comments in the meantime~~

I also updated the load test to hopefully be more consistent with real-world usage by:
1. using a Redis container/cache
2. fixing the context to use a value that would result in a successful evaluation (I think)

### No Cache

```
 [VARIANT]: Vegeta attack for 60.000000 seconds...
 [VARIANT]: Vegeta attack complete

 Authentication Enabled: true
 Cache Enabled: false
 --------------------
 Mean: 4.530004ms
 95th percentile: 5.107738ms
 99th percentile: 5.633436ms
 Max: 20.098ms
 Requests: 18000
 Throughput: 300.007131
 Errors: []
 Status Codes: map[200:18000]
 Success: 100.000000 %


 [BOOLEAN]: Vegeta attack for 60.000000 seconds...
 [BOOLEAN]: Vegeta attack complete

 Authentication Enabled: true
 Cache Enabled: false
 --------------------
 Mean: 1.885374ms
 95th percentile: 2.468148ms
 99th percentile: 3.269184ms
 Max: 55.553791ms
 Requests: 18000
 Throughput: 300.007224
 Errors: []
 Status Codes: map[200:18000]
 Success: 100.000000 %
```

### Cache (Redis Container)

```
 [VARIANT]: Vegeta attack for 60.000000 seconds...
 [VARIANT]: Vegeta attack complete

 Authentication Enabled: true
 Cache Enabled: true
 --------------------
 Mean: 1.861086ms
 95th percentile: 2.370013ms
 99th percentile: 3.15345ms
 Max: 26.475791ms
 Requests: 18000
 Throughput: 300.011936
 Errors: []
 Status Codes: map[200:18000]
 Success: 100.000000 %

 [BOOLEAN]: Vegeta attack complete

 Authentication Enabled: true
 Cache Enabled: true
 --------------------
 Mean: 1.834613ms
 95th percentile: 2.394021ms
 99th percentile: 3.023923ms
 Max: 19.449084ms
 Requests: 18000
 Throughput: 300.009053
 Errors: []
 Status Codes: map[200:18000]
 Success: 100.000000 %
```

### CODE REVIEW

File: build/hack/cmd/loadtest/main.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func main() {
 			EntityId: uuid.Must(uuid.NewV4()).String(),
 			FlagKey:  "flag_010",
 			Context: map[string]string{
-				"in_segment": "baz",
+				"in_segment": "segment_001",
```
[the remark this answers is not in the document]
Comment (markphelps) [reply build/hack/cmd/loadtest/main.go:53]: oh i was getting `no constraint` matched I think in the logs, is there a segment that has this as the last constraint? i see what you're saying that this short circuits it.. lemme compare it with `baz` with cache disabled and see the timings that we get

**Comment by GeorgeMac** [comment]
Approve of the general approach 👍 

And agree r.e. both the measure to handle cache invalidation and prevention, particularly for the UI.
Adding the cache control header makes total sense.

File: internal/cmd/grpc.go [hunk 2]
```diff
@@ -245,7 +245,12 @@ func NewGRPCServer(
 
 	var cacher cache.Cacher
 	if cfg.Cache.Enabled {
-		cacher, cacheShutdown, err := getCache(ctx, cfg)
+		var (
+			cacheShutdown errFunc
+			err           error
+		)
+
+		cacher, cacheShutdown, err = getCache(ctx, cfg)
```
Comment (markphelps) [inline internal/cmd/grpc.go:253]: this was the 🕶️ shadow bug

File: build/testing/integration/api/api.go [hunk 3]
```diff
@@ -230,6 +231,10 @@ func API(t *testing.T, ctx context.Context, client sdk.SDK, namespace string, au
 
 		t.Log("Get flag \"test\" and check variants.")
 
+		// TODO: don't love this..
+		// wait for flag cache to expire when running with cache
+		time.Sleep(750 * time.Millisecond)
```
Comment (GeorgeMac) [inline build/testing/integration/api/api.go:236]: Yeah... I was wondering what the alternative was but the alternatives are dont cache haha.

File: internal/cmd/grpc.go (hunk 2 above)
Comment (GeorgeMac) [reply internal/cmd/grpc.go:253]: classic

**Review by GeorgeMac — APPROVED** [review]
Love it

**Pull request merged**